rate_limit/token_bucket.go

```diff
@@ -0,0 +1,67 @@
+package rate_limit
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+	"time"
+)
+
+type TokenBucket struct {
+	mutex     sync.Mutex
```

Comment:
~use RWMutex instead and use Lock / RLock~
prefer using the `atomic.AddInt64` package instead of mutex